### DISCUSSION TITLE
How to backup UFEI, and BIOS, partitions?

### DISCUSSION THREAD

**Discussion by spflanze** [openzfs/zfs#13786] [Q&A]
I have a working Ubuntu 22.04 installation on a RAIDed encrypted ZFS file system. The procedure I followed can be downloaded from:
https://www.mediafire.com/file/2w8mdb96tbzslub/Ubuntu_22.04_Root_on_ZFS_Encryption.odt/file
This procedure is a customized version of rlaager's at:
https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html

Before I migrate this installation to Pop!_OS I want to be able to easily restore the system to what it was before the Pop!_OS installation attempt. This is necessary because my prior experience is that this installation is a risky thing to do, and  it can make the computer unbootable. The Ubuntu installation procedure requires a lot of time, and effort, that I seek to protect.

The snapshot procedure I am about to do is:
```
# Snapshot the Pools
zfs snapshot bpool/BOOT@snap_boot
zfs snapshot rpool/ROOT@snap_root
zfs snapshot rpool/USERDATA@snap_userdata
zfs snapshot hpool/HOME@snap_home

# Verify the snapshots are created:
zfs list -t snapshot | grep  "@snap"

```
My concern now is about how to backup the UEFI, and BIOS, partitions. What is the best way to do that? How necessary is it to back up these two parttiions to protect against damage a failed Pop!_OS installation attempt might do?

**Comment by m-ueberall** [comment]
Depending on whether the partitions are mounted and only contain files, I either use `tar` or `dd` along with `gpg` to encrypt and archive their contents as far as `zfs send` is not an option. IMHO, being able to restore the contents of these partitions using a recent backup in case of need is easier/faster than to recreate them. (Also, the resulting archives don't need much space at all.)

**Comment by IvanVolosyuk** [comment]
Nice idea with 'dd'. I wonder if there is a way to copy over a partition
image file just the changed blocks, so that I could use snapshots for the
disk image with a minimal overhead.

[update] It looks like I can 'dd' the partition's device to a temporary file and
then 'rsync --inplace' to the destination image :)

[update] Sorry for the conversation with myself, but it looks like rsync --inplace
will still rewrite full file:
$ rsync -av --inplace /var/tmp/uefi-backup.img /archive/replica/boot/
$ zfs list -t snapshot archive/replica/boot
NAME                     USED  AVAIL     REFER  MOUNTPOINT
***@***.***   107M      -      107M  -
***@***.***   127M      -      128M  -
Back to the drawing table or just screw this and accept wasted 100M space.
Is there a tool which will reliably rewrite just changed bytes on the
destination?

**Reply by bghira** [reply]
try `rsync -c` switch so it checksums data.

**Reply by IvanVolosyuk** [reply]
What I read in documentation tells me that --checksum is a per-file
checksum, but not a block-level checksum, thus doesn't help in this case. I
wonder if I can just write a simple .c program which compares and writes
bytes with the specified record-size boundary.

I hacked something really quick :)
https://github.com/IvanVolosyuk/inplace

**Comment by IvanVolosyuk** [comment]
It looks like I can 'dd' the partition's device to a temporary file and
then 'rsync --inplace' to the destination image :)
